### DISCUSSION TITLE
Multiple turbine types, same wake model but with different wake parameters for each turbine type

### DISCUSSION THREAD

**Discussion by cpjordan** [NatLabRockies/floris#845] [Q&A] [answered]
Hello, I see that it is possible to use [multiple turbine types](https://github.com/NREL/floris/pull/325) now, which is fantastic, but I'm wondering whether it is possible to also modify the wake parameters on a per turbine basis? 

I use FLORIS as a bit of a black box, as I use it to [design tidal turbine arrays](https://link.springer.com/article/10.1007/s40722-022-00225-2) at higher speeds than conventional optimisations in this research area. I usually do a calibration exercise for different turbine types and modify e.g. ka, kb, alpha, beta for the Gauss legacy wake model. It would be ideal if I could define multiple turbine types but also the corresponding parameters for each wake. Since for the farm we cycle through each turbine and use its values for diameter, hub height etc., it seems like it should also be feasible to update key wake parameters for the deficit calculation if we keep the same wake model but associate the parameters with each turbine.

I should note that I am using v2.4 for a few reasons, but I can work backwards if there are any examples where this is done! I guess a custom fix for my purposes would be to change the parameters of interest (ka, kb, alpha and beta) to be turbine properties instead? Sorry in advance if I have missed something in the example or in Discussions and this is already possible!

**Comment by rafmudaf** [comment, marked as answer]
Hi @cpjordan No, FLORIS has not yet supported setting wake parameters per turbine. However, the idea has come up a few times in the past. If you were to modify FLORIS to support this, there are two different paths for v2 and v3. In v2, just like you said, it should be a parameter on the Turbine objects. In v3, the wake model parameters should be 3D arrays probably associated to the wake classes that they're already on with an entry for each turbine (i.e. shape would be `(n wind directions, n wind speeds, n turbines)`).

If you do intend to make this change, please engage here. I'm interested in this, as well, and I'm happy to support your efforts. Heads up, though, that v2 is not generally updated these days, and v3 will soon be deprecated. We are releasing the next major update, v4, in the next few weeks.

**Reply by cpjordan** [reply]
Thanks @rafmudaf! This worked exactly as anticipated. For my case, I decided to add the wake parameters I wanted to change as an additional (duplicate) entry into the turbines entry in the input .json. I then make `wake_parameters` an attribute of the `turbines` object, defaulting to None if not provided. In the deficit or deflection calculations, `self.ka`, `self.kb` etc. can then optionally be updated on a per turbine basis if they have been defined. 

I spent a year out of research hence I am behind on the FLORIS version and working with what I had! I have a few other changes I had made so that FLORIS still runs in reasonable time if I have thousands of data points per flow field (for tidal applications). I think this probably conflicts with the general usage, but it would be interesting to see whether some of this stuff could be incorporated in v4 down the line..!
